python/lsst/meas/extensions/shapeHSM/_hsm_moments.py

```diff
@@ -0,0 +1,208 @@
+                record.getId(),
+                center.getX(),
+                center.getY(),
+                error,
+            )
+
+
+class HsmSourceMomentsConfig(HsmMomentsConfig):
+    # later we can put some doscstring
+    badMaskPlanes = pexConfig.ListField[str](
+        doc="Mask planes used to reject bad pixels.", default=["BAD", "SAT"]
+    )
+
+
+@measBase.register("ext_shapeHSM_HsmSourceMoments")
+class HsmSourceMomentsPlugin(HsmMomentsPlugin):
+    ConfigClass = HsmSourceMomentsConfig
+
+    def __init__(self, config, name, schema, metadata, logName=None) -> None:
+        super().__init__(config, name, schema, logName)
+        # some flag thingy
+        # work in conjunction with fail()
+        self.centroidResultKey = afwTable.Point2DKey.addFields(
+            schema, "centroid", "The centroid measured by HSM", "pixel"
+        )
+        self.shapeKey = afwTable.QuadrupoleKey.addFields(schema, name, doc="HSM source moments.")
+        if config.addFlux:  # search the correct attribute name in the codebase
+            self.fluxKey = measBase.FluxResultKey.addFields(schema, "flux", "The flux measured by HSM", "dn")
+
+    def measure(self, record, exposure):
+        """... in place"""
+        # source in c++ is record in py. exposure the same
+
+        center = record.getCentroid()
+        badMaskPlanes = self.config.badMaskPlanes
+        badPixelMask = exposure.getMaskedImage().getMask().getPlaneBitMask(badMaskPlanes)
+        bbox = record.getFootprint().getBBox()
+        if bbox.getArea() == 0:
+            raise MeasurementError("Find the same error message from .cc file")
+
+        if not bbox.contains(center):
+            raise MeasurementError("Find the same error message from .cc file")
+
+        psfSigma = exposure.getPsf().computeShape(center).getTraceRadius()
+
+        # self._calculate(record, exposure, bbox, center, badPixelMask, ...)
+
+        # galsim
+        image_array = exposure[bbox].getImage().array
+        xmin, xmax = bbox.getMinX() + 1, bbox.getMaxX() + 1
+        ymin, ymax = bbox.getMinY() + 1, bbox.getMaxY() + 1
+        bounds = galsim.bounds.BoundsI(xmin, xmax, ymin, ymax)
+        image = galsim.Image(image_array, bounds=bounds, copy=False)
+        # GalSim's HSM uses the FITS convention of 1,1 for the lower-left corner
+        # account for difference in origin while converting center to galsim centerid (guessed)
+        # make sure bbox is centered properly
+
+        try:
+            shape = galsim.hsm.FindAdaptiveMom(
+                image,
+                weight=None,
+                badpix=badPixelMask,
+                guess_sig=2.5 * psfSigma,
+                precision=1.0e-6,
+                guess_centroid=None,
+                strict=True,
+                round_moments=False,
+                hsmparams=None,
+                use_sky_coords=False,
+            )
+
+        except galsim.hsm.GalSimHSMError as e:
+            raise MeasurementError(e)
```

Comment:
```suggestion
            raise MeasurementError(e, self.GALSIM.number)
```
after you transfer the various error flags from the C++ code. Here's a useful document to read: https://github.com/lsst/meas_base/blob/1bfaca56951770c88f4308da41de16d72ce40db9/doc/lsst.meas.base/tasks_and_algorithms.rst#L123

Reply:
Done!